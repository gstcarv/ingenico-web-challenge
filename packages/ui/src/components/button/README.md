# ðŸŽ¯ Button Component

A flexible and accessible button component with multiple variants and sizes.

## ðŸ“¦ Installation

```bash
import { Button } from '@ingenico-challenge/ui'
```

## ðŸš€ Usage

### Basic Button

```tsx
import { Button } from '@ingenico-challenge/ui'

function MyComponent() {
  return (
    <Button onClick={() => console.log('Clicked!')}>
      Click me
    </Button>
  )
}
```

### Button Variants

```tsx
// Primary button (default)
<Button variant="primary">Primary</Button>

// Secondary button
<Button variant="secondary">Secondary</Button>

// Outline button
<Button variant="outline">Outline</Button>

// Ghost button
<Button variant="ghost">Ghost</Button>

// Destructive button
<Button variant="destructive">Delete</Button>
```

### Button Sizes

```tsx
// Small button
<Button size="sm">Small</Button>

// Medium button (default)
<Button size="md">Medium</Button>

// Large button
<Button size="lg">Large</Button>
```

### Disabled State

```tsx
<Button disabled onClick={() => console.log('This won\'t fire')}>
  Disabled Button
</Button>
```

## ðŸ“‹ Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `variant` | `'primary' \| 'secondary' \| 'outline' \| 'ghost' \| 'destructive'` | `'primary'` | Visual style variant |
| `size` | `'sm' \| 'md' \| 'lg'` | `'md'` | Button size |
| `disabled` | `boolean` | `false` | Disables the button |
| `className` | `string` | - | Additional CSS classes |
| `...props` | `ButtonHTMLAttributes` | - | All standard button attributes |

## ðŸŽ¨ Styling

The component uses Class Variance Authority (CVA) for styling variants and supports custom className overrides.

### Base Classes
- Focus-visible ring for accessibility
- Disabled state styling
- Smooth transitions
- Rounded corners

### Variant Styles
- **Primary**: Brand color with shadow effect
- **Secondary**: Neutral background
- **Outline**: Bordered style
- **Ghost**: Minimal styling
- **Destructive**: Error color scheme

## â™¿ Accessibility

- Proper focus management with visible focus ring
- Disabled state handling
- Semantic HTML button element
- Keyboard navigation support

## ðŸ§ª Testing

```tsx
import { render, screen } from '@testing-library/react'
import { Button } from './button'

test('renders button with correct text', () => {
  render(<Button>Test Button</Button>)
  expect(screen.getByRole('button', { name: 'Test Button' })).toBeInTheDocument()
})

test('handles click events', () => {
  const handleClick = jest.fn()
  render(<Button onClick={handleClick}>Click me</Button>)
  
  screen.getByRole('button').click()
  expect(handleClick).toHaveBeenCalledTimes(1)
})
```

## ðŸ“š Storybook

View interactive examples and documentation in Storybook:

```bash
cd packages/ui
pnpm storybook
```

Navigate to the Button component in the Storybook sidebar.
