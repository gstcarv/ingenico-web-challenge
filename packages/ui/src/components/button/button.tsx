import { ComponentProps, forwardRef } from "react";
import { cva, type VariantProps } from "class-variance-authority";
import { cn } from "../../utils";

const buttonVariants = cva(
    "inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer",
    {
        variants: {
            variant: {
                primary:
                    "bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)]",
                secondary: "bg-neutral-100 text-neutral-900 hover:bg-neutral-200 focus-visible:ring-neutral-500",
                outline:
                    "border border-neutral-200 bg-white text-neutral-900 hover:bg-neutral-50 focus-visible:ring-neutral-500",
                ghost: "text-neutral-900 hover:bg-neutral-100 focus-visible:ring-neutral-500",
                destructive: "bg-error-600 text-white hover:bg-error-700 focus-visible:ring-error-500",
            },
            size: {
                sm: "h-8 px-3 text-sm",
                md: "h-10 px-4 text-sm",
                lg: "h-16 px-6 text-base",
            },
        },
        defaultVariants: {
            variant: "primary",
            size: "md",
        },
    },
);

export type ButtonProps = ComponentProps<"button"> & VariantProps<typeof buttonVariants>;

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(({ className, variant, size, ...props }, ref) => {
    return <button className={cn(buttonVariants({ variant, size, className }))} ref={ref} {...props} />;
});

Button.displayName = "Button";
