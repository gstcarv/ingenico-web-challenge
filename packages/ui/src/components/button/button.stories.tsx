import type { Meta, StoryObj } from "@storybook/react-vite";
import { Button } from "./button";
import { Download, Heart, Plus } from "lucide-react";

const meta: Meta<typeof Button> = {
    title: "Components/Button",
    component: Button,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
    argTypes: {
        variant: {
            control: "select",
            options: ["primary", "secondary", "outline", "ghost", "destructive"],
        },
        size: {
            control: "select",
            options: ["sm", "md", "lg"],
        },
        disabled: {
            control: "boolean",
        },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        children: "Button",
    },
};

export const Primary: Story = {
    args: {
        variant: "primary",
        children: "Primary Button",
    },
};

export const Secondary: Story = {
    args: {
        variant: "secondary",
        children: "Secondary Button",
    },
};

export const Outline: Story = {
    args: {
        variant: "outline",
        children: "Outline Button",
    },
};

export const Ghost: Story = {
    args: {
        variant: "ghost",
        children: "Ghost Button",
    },
};

export const Destructive: Story = {
    args: {
        variant: "destructive",
        children: "Destructive Button",
    },
};

export const Small: Story = {
    args: {
        size: "sm",
        children: "Small Button",
    },
};

export const Large: Story = {
    args: {
        size: "lg",
        children: "Large Button",
    },
};

export const Disabled: Story = {
    args: {
        disabled: true,
        children: "Disabled Button",
    },
};

export const WithIcon: Story = {
    args: {
        children: (
            <>
                <Plus className="mr-2 h-4 w-4" />
                Add Item
            </>
        ),
    },
};

export const IconOnly: Story = {
    args: {
        children: <Heart className="h-4 w-4" />,
        "aria-label": "Like",
    },
};

export const WithIconAfter: Story = {
    args: {
        children: (
            <>
                Download
                <Download className="ml-2 h-4 w-4" />
            </>
        ),
    },
};

export const AllVariants: Story = {
    render: () => (
        <div className="flex flex-wrap gap-4">
            <Button variant="primary">Primary</Button>
            <Button variant="secondary">Secondary</Button>
            <Button variant="outline">Outline</Button>
            <Button variant="ghost">Ghost</Button>
            <Button variant="destructive">Destructive</Button>
        </div>
    ),
};

export const AllSizes: Story = {
    render: () => (
        <div className="flex items-center gap-4">
            <Button size="sm">Small</Button>
            <Button size="md">Medium</Button>
            <Button size="lg">Large</Button>
        </div>
    ),
};
