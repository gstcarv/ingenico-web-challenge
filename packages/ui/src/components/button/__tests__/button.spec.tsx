import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { Button } from "../button";

describe("Button", () => {
    it("renders button element with correct attributes", () => {
        render(<Button data-testid="button">Click me</Button>);

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toBeInTheDocument();
        expect(buttonElement.tagName).toBe("BUTTON");
        expect(buttonElement).toHaveTextContent("Click me");
    });

    it("applies custom className", () => {
        const { container } = render(
            <Button className="custom-class" data-testid="button">
                Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("handles click events", () => {
        const handleClick = vi.fn();

        render(
            <Button onClick={handleClick} data-testid="button">
                Click me
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        fireEvent.click(buttonElement);

        expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it("handles disabled state", () => {
        const handleClick = vi.fn();

        render(
            <Button disabled onClick={handleClick} data-testid="button">
                Disabled Button
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toBeDisabled();

        fireEvent.click(buttonElement);
        expect(handleClick).not.toHaveBeenCalled();
    });

    it("forwards ref correctly", () => {
        const ref = vi.fn();

        render(
            <Button ref={ref} data-testid="button">
                Button
            </Button>,
        );

        expect(ref).toHaveBeenCalledWith(screen.getByTestId("button"));
    });

    it("forwards all HTML button attributes", () => {
        render(
            <Button data-testid="button" name="submit-button" id="submit-btn" type="submit" aria-label="Submit form">
                Submit
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toHaveAttribute("name", "submit-button");
        expect(buttonElement).toHaveAttribute("id", "submit-btn");
        expect(buttonElement).toHaveAttribute("type", "submit");
        expect(buttonElement).toHaveAttribute("aria-label", "Submit form");
    });
});

describe("Button Variants", () => {
    it("applies primary variant classes", () => {
        const { container } = render(
            <Button variant="primary" data-testid="button">
                Primary Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies secondary variant classes", () => {
        const { container } = render(
            <Button variant="secondary" data-testid="button">
                Secondary Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies outline variant classes", () => {
        const { container } = render(
            <Button variant="outline" data-testid="button">
                Outline Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies ghost variant classes", () => {
        const { container } = render(
            <Button variant="ghost" data-testid="button">
                Ghost Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies destructive variant classes", () => {
        const { container } = render(
            <Button variant="destructive" data-testid="button">
                Destructive Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("uses default variant when not specified", () => {
        const { container } = render(<Button data-testid="button">Default Button</Button>);

        expect(container.firstChild).toMatchSnapshot();
    });
});

describe("Button Sizes", () => {
    it("applies small size classes", () => {
        const { container } = render(
            <Button size="sm" data-testid="button">
                Small Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies medium size classes", () => {
        const { container } = render(
            <Button size="md" data-testid="button">
                Medium Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("applies large size classes", () => {
        const { container } = render(
            <Button size="lg" data-testid="button">
                Large Button
            </Button>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it("uses default size when not specified", () => {
        const { container } = render(<Button data-testid="button">Default Size Button</Button>);

        expect(container.firstChild).toMatchSnapshot();
    });
});

describe("Button with Icons", () => {
    it("renders button with icon before text", () => {
        const TestIcon = () => <span data-testid="icon">🚀</span>;

        render(
            <Button data-testid="button">
                <TestIcon />
                Launch
            </Button>,
        );

        expect(screen.getByTestId("icon")).toBeInTheDocument();
        expect(screen.getByTestId("button")).toHaveTextContent("Launch");
    });

    it("renders button with icon after text", () => {
        const TestIcon = () => <span data-testid="icon">📤</span>;

        render(
            <Button data-testid="button">
                Send
                <TestIcon />
            </Button>,
        );

        expect(screen.getByTestId("icon")).toBeInTheDocument();
        expect(screen.getByTestId("button")).toHaveTextContent("Send");
    });

    it("renders icon-only button", () => {
        const TestIcon = () => <span data-testid="icon">❤️</span>;

        render(
            <Button data-testid="button" aria-label="Like">
                <TestIcon />
            </Button>,
        );

        expect(screen.getByTestId("icon")).toBeInTheDocument();
        expect(screen.getByTestId("button")).toHaveAttribute("aria-label", "Like");
    });
});

describe("Button Accessibility", () => {
    it("supports aria-label", () => {
        render(
            <Button aria-label="Submit form" data-testid="button">
                Submit
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toHaveAttribute("aria-label", "Submit form");
    });

    it("supports aria-describedby", () => {
        render(
            <Button aria-describedby="description" data-testid="button">
                Button
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toHaveAttribute("aria-describedby", "description");
    });

    it("supports role attribute", () => {
        render(
            <Button role="menuitem" data-testid="button">
                Menu Item
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toHaveAttribute("role", "menuitem");
    });

    it("has proper focus styles", () => {
        render(<Button data-testid="button">Focusable Button</Button>);

        const buttonElement = screen.getByTestId("button");
        buttonElement.focus();
        expect(buttonElement).toHaveFocus();
    });
});

describe("Button Integration", () => {
    it("works with form elements", () => {
        const handleSubmit = vi.fn((e) => e.preventDefault());

        render(
            <form onSubmit={handleSubmit}>
                <Button type="submit" data-testid="button">
                    Submit Form
                </Button>
            </form>,
        );

        const buttonElement = screen.getByTestId("button");
        fireEvent.click(buttonElement);

        expect(handleSubmit).toHaveBeenCalledTimes(1);
    });

    it("can be focused and handles click events", () => {
        const handleClick = vi.fn();

        render(
            <Button onClick={handleClick} data-testid="button">
                Focusable Button
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");

        // Focus the button
        buttonElement.focus();
        expect(buttonElement).toHaveFocus();

        // Click the button
        fireEvent.click(buttonElement);
        expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it("prevents click when disabled", () => {
        const handleClick = vi.fn();

        render(
            <Button disabled onClick={handleClick} data-testid="button">
                Disabled Button
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");

        // Try clicking
        fireEvent.click(buttonElement);
        expect(handleClick).not.toHaveBeenCalled();

        // Try keyboard interaction
        fireEvent.keyDown(buttonElement, { key: "Enter" });
        expect(handleClick).not.toHaveBeenCalled();
    });
});

describe("Button Edge Cases", () => {
    it("handles empty children", () => {
        render(<Button data-testid="button" />);

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toBeInTheDocument();
        expect(buttonElement).toHaveTextContent("");
    });

    it("handles null children", () => {
        render(<Button data-testid="button">{null}</Button>);

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toBeInTheDocument();
        expect(buttonElement).toHaveTextContent("");
    });

    it("handles undefined children", () => {
        render(<Button data-testid="button">{undefined}</Button>);

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toBeInTheDocument();
        expect(buttonElement).toHaveTextContent("");
    });

    it("handles complex children", () => {
        render(
            <Button data-testid="button">
                <span>Text</span>
                <strong>Bold</strong>
                <em>Italic</em>
            </Button>,
        );

        const buttonElement = screen.getByTestId("button");
        expect(buttonElement).toHaveTextContent("TextBoldItalic");
    });
});
