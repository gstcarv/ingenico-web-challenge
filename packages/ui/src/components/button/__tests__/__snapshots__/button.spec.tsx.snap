// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > applies custom className 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-10 px-4 text-sm custom-class"
  data-testid="button"
>
  Button
</button>
`;

exports[`Button Sizes > applies large size classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-16 px-6 text-base"
  data-testid="button"
>
  Large Button
</button>
`;

exports[`Button Sizes > applies medium size classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-10 px-4 text-sm"
  data-testid="button"
>
  Medium Button
</button>
`;

exports[`Button Sizes > applies small size classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-8 px-3 text-sm"
  data-testid="button"
>
  Small Button
</button>
`;

exports[`Button Sizes > uses default size when not specified 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-10 px-4 text-sm"
  data-testid="button"
>
  Default Size Button
</button>
`;

exports[`Button Variants > applies destructive variant classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-error-600 text-white hover:bg-error-700 focus-visible:ring-error-500 h-10 px-4 text-sm"
  data-testid="button"
>
  Destructive Button
</button>
`;

exports[`Button Variants > applies ghost variant classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer text-neutral-900 hover:bg-neutral-100 focus-visible:ring-neutral-500 h-10 px-4 text-sm"
  data-testid="button"
>
  Ghost Button
</button>
`;

exports[`Button Variants > applies outline variant classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer border border-neutral-200 bg-white text-neutral-900 hover:bg-neutral-50 focus-visible:ring-neutral-500 h-10 px-4 text-sm"
  data-testid="button"
>
  Outline Button
</button>
`;

exports[`Button Variants > applies primary variant classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-10 px-4 text-sm"
  data-testid="button"
>
  Primary Button
</button>
`;

exports[`Button Variants > applies secondary variant classes 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-neutral-100 text-neutral-900 hover:bg-neutral-200 focus-visible:ring-neutral-500 h-10 px-4 text-sm"
  data-testid="button"
>
  Secondary Button
</button>
`;

exports[`Button Variants > uses default variant when not specified 1`] = `
<button
  class="inline-flex items-center justify-center rounded-lg font-bold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 cursor-pointer bg-brand-primary text-white hover:bg-brand-600 focus-visible:ring-brand-primary shadow-[0_4px_12px_rgba(25,182,221,0.3)] h-10 px-4 text-sm"
  data-testid="button"
>
  Default Button
</button>
`;
